### ISSUE TITLE
Stack overflow during type inference

### ISSUE THREAD

**Issue by vchuravy** [JuliaLang/julia#55102]
```
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.4 (2024-06-04)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> f(::Val{1000}) = 1000
f (generic function with 1 method)

julia> f(::Val{N}) where N = f(Val(N-1))
f (generic function with 2 methods)

julia> f(Val(2000))
Internal error: stack overflow in type inference of f(Base.Val{2000}).
This might be caused by recursion over very long tuples or argument lists.
1000
```

In `1.9` this didn't print an error.

**Comment by christiangnrd** [comment]
This seems to only error on x86. It errors on x86 Windows and Linux , but not ARM MacOS or Linux

**Comment by oscardssmith** [comment]
wait, that's way more cursed. how is type inference platform dependent?

**Comment by KristofferC** [comment]
Size of stack frames and how big it can be before overflowing is platform dependent, no?

**Comment by christiangnrd** [comment]
Ok yes calling `f` with a higher values also errors on my ARM devices.

**Comment by christiangnrd** [comment]
Bisected to #48913.

**Comment by bvdmitri** [comment]
I've encountered this error while debugging `@testitem` in the preview version of VSCode. 

<img width="555" alt="image" src="https://github.com/user-attachments/assets/22f72fb2-5307-49a9-b5b0-8a70e26e6b4b">

The test process crashed immediately after clicking the button with

```julia
StackOverflowError()
Internal error: stack overflow in type inference of handle_err(Any, TestItemServer.JuliaInterpreter.Frame, StackOverflowError).
This might be caused by recursion over very long tuples or argument lists.
```

**Comment by nsajko** [comment]
@bvdmitri the cause behind your stack overflow in type inference could be different, though? If you're looking for a quick fix, try increasing the stack size, xref #54998

**Comment by bvdmitri** [comment]
Well, I have no idea what is the cause. The error clearly says that the problem is related to `TestItemServer.JuliaInterpreter.Frame` and the error message is the same as in this topic. It just crashed the VSCode extension when I tried to debug one of the `@testitems`. I just thought it would make this issue more relevant

**Comment by nsajko** [comment]
Fixed by #55575
